File: src/main/java/schemabuilder/processor/pipelines/parsing/datafetchers/GraphQLTypeParser.java
package schemabuilder.processor.pipelines.parsing.datafetchers;

import graphql.schema.DataFetcher;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import schemabuilder.annotations.graphql.GraphQLDataFetcher;
import schemabuilder.annotations.graphql.GraphQLTypeConfiguration;
import schemabuilder.processor.pipelines.parsing.GraphQLClassParserStrategy;
import schemabuilder.processor.wiring.InstanceFetcher;

public class GraphQLTypeParser implements GraphQLClassParserStrategy {

    private GraphQLDataFetcherBank dataFetcherBank = GraphQLDataFetcherBank.getInstance();

    @Override
    public void parse(Class<?> clazz, InstanceFetcher fetcher) {
        if (!clazz.isAnnotationPresent(GraphQLTypeConfiguration.class)) {
            return;
        }

        String typeName = clazz.getAnnotation(GraphQLTypeConfiguration.class).value();
        Object instance = fetcher.getInstance(clazz);


        for (Method method : clazz.getDeclaredMethods()) {
            GraphQLDataFetcher annotation = method.getAnnotation(GraphQLDataFetcher.class);
            if(annotation == null || !method.getReturnType().equals(DataFetcher.class)) {
                continue;
            }

            String fieldName;
            if(annotation.value().equals("")) {
                fieldName = method.getName();
            } else {
                fieldName = annotation.value();
            }

            int cost = annotation.cost();

            method.setAccessible(true);
            try {
                dataFetcherBank.addDataFetcher(new GraphQLDataFetcherType(typeName, cost, fieldName, (DataFetcher<?>) method.invoke(instance)));
            } catch (IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }
        }
    }
}

File: src/main/java/schemabuilder/processor/pipelines/parsing/datafetchers/GraphQLTypeParser.kt
package schemabuilder.processor.pipelines.parsing.datafetchers

import graphql.schema.DataFetcher
import graphql.schema.DataFetchingEnvironment
import schemabuilder.annotations.graphql.GraphQLDataFetcher
import schemabuilder.annotations.graphql.GraphQLTypeConfiguration
import schemabuilder.processor.pipelines.parsing.GraphQLClassParserStrategy
import schemabuilder.processor.pipelines.parsing.ParsedResults
import schemabuilder.processor.wiring.InstanceFetcher
import kotlin.reflect.KClass
import kotlin.reflect.full.declaredFunctions
import kotlin.reflect.full.findAnnotation
import kotlin.reflect.full.hasAnnotation
import kotlin.reflect.full.isSubtypeOf
import kotlin.reflect.jvm.isAccessible
import kotlin.reflect.typeOf

class GraphQLTypeParser : GraphQLClassParserStrategy {

    @ExperimentalStdlibApi
    override fun parse(clazz: KClass<*>, fetcher: InstanceFetcher) {
        val typeName: String = clazz.findAnnotation<GraphQLTypeConfiguration>()?.value!!

        for (method in clazz.declaredFunctions) {
            val annotation = method.findAnnotation<GraphQLDataFetcher>()!!

            val fieldName = if (annotation.value == "") {
                method.name
            } else {
                annotation.value
            }

            val cost: Int = annotation.cost
            method.isAccessible = true

            try {
                val dataFetcherInstance = DataFetcher { environment ->
                    val paramNames = method.parameters.map { Pair(it.name, it.type) }.filter { it.first != null }

                    val params: MutableList<Any?> = paramNames.map { (first, second) ->
                        return@map if (second.isSubtypeOf(typeOf<DataFetchingEnvironment>())) {
                            environment
                        } else {
                            environment.arguments[first]
                        }
                    }.toMutableList()

                    params.add(0, fetcher.getInstance(clazz))

                    method.call(*params.toTypedArray())
                }

                DataFetcherCostMap.setCostFor(dataFetcherInstance, cost)
                ParsedResults.datafetchers.add(GraphQLDataFetcherType(typeName, cost, fieldName, dataFetcherInstance))
            } catch (e: Exception) { e.printStackTrace() }
        }
    }

    @ExperimentalStdlibApi
    override fun shouldParse(clazz: KClass<*>) = clazz.hasAnnotation<GraphQLTypeConfiguration>()
}